File: exercism/python/simple-cipher/cipher.py
from functools import partial
from itertools import cycle, starmap
from random import randint

orda = ord('a')
ordz = ord('z')


def shift(x, y):
    return chr(orda + (ord(x) + ord(y) - 2 * orda) % 26)


def unshift(x, y):
    return chr(orda + (ord(x) - ord(y)) % 26)


def random_key(length):
    return ''.join(chr(randint(orda, ordz)) for _ in range(100))


class Cipher(object):
    def __init__(self, key=None):
        if key and any(not ch.isalpha() or not ch.islower() for ch in key):
            raise ValueError

        self.key = key or random_key(length=100)
        self.encode = partial(self._translate, shift)
        self.decode = partial(self._translate, unshift)

    def _translate(self, f, given):
        chars = (ch.lower() for ch in given if ch.isalpha())
        return ''.join(starmap(f, zip(chars, cycle(self.key))))


class Caesar(Cipher):
    def __init__(self):
        super(Caesar, self).__init__('d')

File: exercism/csharp/simple-cipher/Cipher.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Cipher
{
    public string Key;

    public Cipher ()
    {
        var r = new Random(Guid.NewGuid().GetHashCode());
        var cipherChars = new int[100].Select(_ => (char)('a' + r.Next(0, 26)));
        Key = String.Concat(cipherChars);
    }

    public Cipher (string key)
    {
        if (key.Length == 0 ||
            key.Where(ch => !Char.IsLower(ch) || !Char.IsLetter(ch)).Any()) {
            throw new ArgumentException();
        }
        Key = key;
    }

    public IEnumerable<char> Encode (IEnumerable<char> given)
    {
        return given.Select((ch, i) => (char)('a' + (ch + Key[i % Key.Length] - 'a' - 'a') % 26));
    }

    public IEnumerable<char> Decode (IEnumerable<char> given)
    {
        return given.Zip(Key, (a, b) => (char)((26 + a - b) % 26 + 'a'));
    }
}
